kyo-data/shared/src/main/scala/kyo/KArray.scala

```diff
@@ -0,0 +1,350 @@
+package kyo
+
+import java.util.Arrays
+import scala.annotation.tailrec
+import scala.collection.immutable.ArraySeq
+import scala.reflect.ClassTag
+
+/** An efficient, immutable array-backed sequence of elements.
+  *
+  * KArray is similar to Scala's built-in IArray (immutable array), but provides optimized methods for common operations and additional
+  * utility functions. It offers better performance characteristics than standard Scala collections for many use cases while maintaining
+  * immutability semantics. In practice, using KArray is akin to using Array directly but with a richer and safer immutable API.
+  *
+  * KArray is essentially a thin wrapper around arrays, making it ideal for scenarios where you need predictable, array-like performance
+  * with O(1) indexing and iteration. It excels in memory-sensitive contexts due to its minimal overhead.
+  *
+  * A key advantage of KArray over Chunk is that KArray doesn't box primitive types, while Chunk does. This makes KArray more
+  * memory-efficient when working with primitive values like Int, Long, Double, etc.
+  *
+  * Important: Unlike Chunk, KArray is NOT part of Scala's collection hierarchy. It doesn't extend Seq or any other collection trait, which
+  * means it won't work with methods expecting standard Scala collections. Consider using Chunk instead when you need compatibility with
+  * Scala's collection library or when you require efficient structural operations like slicing without copying data.
+  *
+  * @tparam A
+  *   the type of elements in this KArray
+  */
+opaque type KArray[A] = Array[A]
```

Comment:
```suggestion
opaque type KArray[+A] = Array[A]
```
Could this be covariant since the array an be stored to?

Reply:
Good point. I wonder if we could have trouble with the fact that the arrays are specialized but it seems worth exploring. Do you mind creating an issue for it?

Reply:
Done
 * https://github.com/getkyo/kyo/issues/1185